components/snap_recovery/src/data_resolver.rs

```diff
@@ -0,0 +1,459 @@
+    tx: UnboundedSender<ResolveKvDataResponse>,
+}
+
+#[allow(dead_code)]
+impl LockResolverWorker {
+    pub fn new(
+        mut lock_iter: RocksEngineIterator,
+        tx: UnboundedSender<ResolveKvDataResponse>,
+    ) -> Self {
+        lock_iter.seek_to_first().unwrap();
+        Self { lock_iter, tx }
+    }
+    pub fn cleanup_lock(&mut self, wb: &mut RocksWriteBatchVec) -> Result<bool> {
+        let mut key_count: u64 = 0;
+        while self.lock_iter.valid().unwrap() {
+            box_try!(wb.delete_cf(CF_LOCK, self.lock_iter.key()));
+            self.lock_iter.next().unwrap();
+            key_count += 1;
+        }
+        info!("clean up lock cf. delete key count {}", key_count);
+        let mut write_opts = WriteOptions::new();
+        write_opts.set_sync(true);
+        if let Err(e) = wb.write_opt(&write_opts) {
+            panic!(
+                "snapshot recovery, fail to write to disk while delete lock, the error is {:?}",
+                e,
+            );
+        }
+        let mut response = ResolveKvDataResponse::default();
+
+        response.set_resolved_key_count(key_count);
+        if let Err(e) = self.tx.unbounded_send(response) {
+            warn!("send the cleanup lock key failure {}", e);
+            if e.is_disconnected() {
+                warn!("channel is disconnected.");
+                return Ok(false);
+            }
+        }
+        Ok(true)
+    }
+}
+
+// TODO: as we tested, this size may more effective than set to 256 (max write
+// batch) a more robust test need to figure out what is best.
+const BATCH_SIZE_LIMIT: usize = 1024 * 1024;
+/// `WriteResolverWorker` is the worker that does the actual delete data work.
+pub struct WriteResolverWorker {
+    batch_size_limit: usize,
+    /// `resolved_ts` is the timestamp to data delete to.
+    resolved_ts: TimeStamp,
+    write_iter: RocksEngineIterator,
+    /// send progress of this task
+    tx: UnboundedSender<ResolveKvDataResponse>,
+}
+
+/// `Batch` means a batch of writes load from the engine.
+/// We scan writes in batches to prevent huge memory usage.
+struct Batch {
+    writes: Vec<(Vec<u8>, Write)>,
+    has_more: bool,
+}
+
+#[allow(dead_code)]
+impl WriteResolverWorker {
+    pub fn new(
+        mut write_iter: RocksEngineIterator,
+        resolved_ts: TimeStamp,
+        tx: UnboundedSender<ResolveKvDataResponse>,
+    ) -> Self {
+        write_iter.seek_to_first().unwrap();
+        Self {
+            batch_size_limit: BATCH_SIZE_LIMIT,
+            write_iter,
+            resolved_ts,
+            tx,
+        }
+    }
+    pub fn resolve_write(&mut self, wb: &mut RocksWriteBatchVec) {
+        let now = Instant::now();
+        while self
+            .batch_resolve_write(wb)
+            .expect("resolve kv data failed when deleting in an invalid cf")
+        {}
+        info!("resolve write takes {}", now.elapsed().as_secs());
+    }
+
+    fn next_batch(&mut self) -> Result<Option<(Vec<u8>, Write)>> {
```

Comment:
It emits a sole key, maybe rename it with `next_write`.

Reply:
fixed